ocaml/typing/mode_intf.mli

```diff
@@ -23,6 +23,10 @@ module type Lattice = sig
 
   val legacy : t
 
+  val less_or_equal : t -> t -> Misc.Le_result.t
+
+  val equal : t -> t -> bool
+
```

Comment:
I agree that `eq` and `le` (together) are equivalent in expressiveness to `less_or_equal`. An alternative implementation could make a `Minimal_lattice` module type without `equal` and `le`, a `Lattice` module type with them, and a functor `Make_lattice (L : Minimal_lattice) : Lattice with type t = L.t` that derives them. Because they're so simple, though, I thought this extra structure wasn't worth it. Or we could drop `equal` and `le` and just use `less_or_equal` everywhere, but that interface is harder to work with.

In terms of usefulness: the kind infrastructure wants to know the difference between "less than" and "equal", because it helps guide error messages. Specifically, when we compute the meet between `m1` and `m2`, if they're equal, the operation is uninteresting (and should not change error messages). If, on the other hand, the new mode is really less than the old one, we've learned something materially new, and we want to update error messages accordingly.

For `mode_cross_left`, we could actually just use `le` -- that one is not a problem.

Reply:
For kinds, I care only about constants. I don't think we'll need mode variables in kinds.

And, yes, it's true that `le` is sufficient, as you say. But I think it's more efficient to implement `le` in terms of `less_or_equal` than the other way around. In particular, I think if we inline `less_or_equal` (and `Le_result.is_le`) into `le`, we'll get the same code as now, but I don't see how the optimizer could do something similar with two calls to `le`.

Reply:
>For kinds, I care only about constants. I don't think we'll need mode variables in kinds.

Maybe I misunderstood - I thought in your original message, you meant that the kind system wants to know if mode crossing `m_new = meet m b`indeed has changed the mode `m` of some value. In that case, the `m` could often be a mode variable.

I believe `Le_result` is used in both the kind and mode system. I'm not sure about kinds. But I think for modes, `le` alone is mostly enough, because:
1. Most of time, we really just want to know if `a <= b`, instead of `a = b`.
2. In the rare cases where we do want to know `a = b`, highly likely that we already know `a <= b`, and it's sufficient to just test `b <= a` to know `a = b`.

In both cases, it's more efficient to run `le` directly instead of going through `less_or_equal`.

Going back to this particular PR, I believe `Mode.equal` is used only for `mode_cross_left` and `expect_mode_cross`, both can be adapted to use `Mode.le` instead. For example:
```
Uniqueness.Const.equal upper_bounds.uniqueness Uniqueness.Const.min
```
can be written as 
```
Uniqueness.Const.le upper_bounds.uniqueness Uniqueness.Const.min
```

This is how `solver.ml` test something for min/max and optimise accordingly.

Reply:
> I'm not sure about kinds.

It's true that `le` is sufficient for modes. But it's not sufficient for kinds.

`Jkind.combine_histories` uses `Desc.sub` (and cares whether `Desc.sub` returns `Less` or `Equal`). `Desc.sub` calls `Const.sub`, which calls `Mode.Alloc.Const.less_or_equal` -- still caring about the difference between `Less` and `Equal`. That's why I added the new functionality.

Reply:
Ahh I see. Do you think it might be better to define inside `jkind.ml`:
```
val alloc_const_less_or_equal : ...
```
who is defined based on `Alloc.Const.le a b` and `Alloc.Const.le b a`?

Reasons:
- Adding `less_or_equal` to all axes adds lots of code, which becomes heavier as we add more axes. 
- `less_or_equal` is not a canonical interface of a lattice (`le` is), and used only by `jkind`.
- Is this `alloc_const_less_or_equal` slower? I think "two small pattern match vs. a single big pattern match" is hard to predict.
- Is `le` going through `less_or_equal` slower than a direct `le`? I thought inlining would make then same; but actually not - see my thread in the internal channel. That means the solver would be made slower.

Reply:
I don't see why `less_or_equal` is any more or less canonical than `le`. Each can be implemented in terms of the other. Mathematicians are more likely to think about `le`, but I don't think that matters very much here.

The most performant thing will be to write all of these functions directly, not in terms of one another.

But I agree that this is not a big concern.

I think mostly this doesn't matter. And so that means the best thing is to stop debating. So I'll just do what you ask.